actors/builtin/miner/miner_actor.go

```diff
@@ -228,6 +229,46 @@ func (a Actor) ConfirmUpdateWorkerKey(rt Runtime, params *abi.EmptyValue) *abi.E
 	return nil
 }
 
+// Proposes or confirms a change of owner address.
+// If invoked by the current owner, proposes a new owner address for confirmation. If the proposed address is the
+// current owner address, revokes any existing proposal.
+// If invoked by and re-proposing the previously proposed address, changes the current owner address to be
+// that proposed address.
+func (a Actor) ChangeOwnerAddress(rt Runtime, newAddress *addr.Address) *abi.EmptyValue {
+	if newAddress.Empty() {
+		rt.Abortf(exitcode.ErrIllegalArgument, "empty address")
+	}
+	if newAddress.Protocol() != addr.ID {
+		rt.Abortf(exitcode.ErrIllegalArgument, "owner address must be an ID address")
+	}
+	var st State
+	rt.StateTransaction(&st, func() {
+		info := getMinerInfo(rt, &st)
+		if rt.Caller() == info.Owner || info.PendingOwnerAddress == nil {
```

Comment:
This looks correct. I had to make the table below to convince myself all the cases were covered. 
```
POA==nil && Owner && NA==Owner => POA=nil
POA==nil && Owner && NA!=Owner => POA=NA
POA==nil && !Owner => Not Owner Error
POA!=nil && Owner => POA=NA
POA!=nil && !Owner && !POA => Not POA Error
POA!=nil && !Owner && POA && NA!=POA => Error
POA!=nil && !Owner && POA && NA==POA => Owner=POA, POA=nil
```
An alternate construction:
```
if Owner:
   validate caller is owner
   info.PendingOwnerAddress = newAddress
else:
  validate PendingOwnerAddress is not nil # if needed
  validate caller is PendingOwnerAddress
  validate newAddress == PendingOwnerAddress
  info.Owner = PendingOwnerAddress

if PendingOwnerAddress == info.Owner:
  info.PendingOwnerAddress == nil
```
This separates the authorization logic from the logic that determines whether this has a future change or not and seems a little more clear to me, but it's up to you.

Reply:
I mostly achieved your suggestion. One constraint is my goal to fail via `ValidateImmediateCallerIs` (hence `SysErrForbidden`) in all paths except confirming the wrong address. So if `info.PendingOwnerAddress == nil`, via `ValidateImmediateCallerIs(info.Owner)`.